### PR TITLE
PMM-6129 Basic API.

### PR DESCRIPTION
**Pull request by BupycHuk**

### CODE REVIEW

File: cmd/dbaas-controller/main.go [hunk 1]
```diff
@@ -32,9 +31,9 @@ import (
 )
 
 func main() {
-	if version.Version == "" {
-		panic("dbaas-controller version is not set during build.")
-	}
+	//if version.Version == "" {
+	//	panic("dbaas-controller version is not set during build.")
+	//}
```
Comment (AlekSi) [inline cmd/dbaas-controller/main.go:36]: Why? That should be set by Makefile now, no?

File: utils/logger/logrus_logger.go [hunk 2]
```diff
@@ -48,6 +48,7 @@ func SetupGlobal() {
 			return
 		},
 	})
+	logrus.SetLevel(logrus.DebugLevel)
```
Comment (AlekSi) [inline utils/logger/logrus_logger.go:51]: I think we should handle `PMM_DEBUG` / `PMM_TRACE` there

File: service/cluster/xtra_db_cluster.go [hunk 3]
```diff
@@ -29,6 +29,15 @@ import (
 	"github.com/percona-platform/dbaas-controller/utils/logger"
 )
 
+// pxcStatesMap matches pxc app states to cluster states.
+var pxcStatesMap = map[k8sclient.ClusterState]controllerv1beta1.XtraDBClusterState{
+	k8sclient.ClusterStateInvalid:  controllerv1beta1.XtraDBClusterState_XTRA_DB_CLUSTER_STATE_INVALID,
+	k8sclient.ClusterStateChanging: controllerv1beta1.XtraDBClusterState_XTRA_DB_CLUSTER_STATE_CHANGING,
+	k8sclient.ClusterStateReady:    controllerv1beta1.XtraDBClusterState_XTRA_DB_CLUSTER_STATE_READY,
+	k8sclient.ClusterStateFailed:   controllerv1beta1.XtraDBClusterState_XTRA_DB_CLUSTER_STATE_FAILED,
+	k8sclient.ClusterStateDeleting: 4, // TODO: fix it
```
[the remark this answers is not in the document]
Comment (AlekSi) [reply service/cluster/xtra_db_cluster.go:38]: https://github.com/percona-platform/dbaas-api/pull/21

[the remark this answers is not in the document]
Comment (BupycHuk) [reply service/cluster/xtra_db_cluster.go:38]: Thanks

File: cmd/dbaas-controller/main.go (hunk 1 above)
Comment (BupycHuk) [reply cmd/dbaas-controller/main.go:36]: Committed accidentally.
It doesn't set ldflags when we call `make run`

File: utils/logger/logrus_logger.go (hunk 2 above)
Comment (BupycHuk) [reply utils/logger/logrus_logger.go:51]: Yes, but I set it for now to have the same level in zap and in logrus.

File: cmd/dbaas-controller/main.go (hunk 1 above)
Comment (AlekSi) [reply cmd/dbaas-controller/main.go:36]: Please update `install` target to set them then

File: utils/logger/logrus_logger.go (hunk 2 above)
Comment (AlekSi) [reply utils/logger/logrus_logger.go:51]: Actually, the default value should be info – and on that level we should log too much, and should not log sensitive details. Both `PMM_DEBUG` and `PMM_TRACE` should enable debug level (and development mode in zap); logrus trace level should not be used

**Comment by AlekSi** [comment]
Please don't forget about that bit from the Confluence:
> dbaas-controller returns FailedPrecondition error for API calls that modify CR that is not in a finished state, preventing concurrent updates.

File: service/cluster/xtra_db_cluster.go [hunk 4]
```diff
@@ -44,23 +53,119 @@ func (s *Service) ListXtraDBClusters(ctx context.Context, req *controllerv1beta1
...
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return new(controllerv1beta1.CreateXtraDBClusterResponse), nil
 }
 
 // UpdateXtraDBCluster updates existing XtraDB cluster.
 func (s *Service) UpdateXtraDBCluster(ctx context.Context, req *controllerv1beta1.UpdateXtraDBClusterRequest) (*controllerv1beta1.UpdateXtraDBClusterResponse, error) {
-	return nil, status.Error(codes.Unimplemented, s.p.Sprintf("This method is not implemented yet."))
+	client := k8sclient.NewK8Client(logger.Get(ctx))
```
Comment (AlekSi) [inline service/cluster/xtra_db_cluster.go:129]: That API is not in scope of that ticket.

File: tests/init.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		if ok {
+			err := f.Value.Set(env)
+			if err != nil {
+				logrus.Fatalf("Invalid ENV variable %s: %s", envVar, env)
+			}
+		}
+	}
+
+	if *debugF {
+		logrus.SetLevel(logrus.DebugLevel)
```
Comment (BupycHuk) [inline tests/init.go:73]: should we reuse logger package from the same repository or let logrus be logger for tests?

**Comment by AlekSi** [comment]
CI fails on "Generate code" step

File: service/cluster/xtra_db_cluster.go [hunk 6]
```diff
@@ -28,6 +28,15 @@ import (
 	"github.com/percona-platform/dbaas-controller/service/k8sclient"
 )
 
+// pxcStatesMap matches pxc app states to cluster states.
+var pxcStatesMap = map[k8sclient.ClusterState]controllerv1beta1.XtraDBClusterState{
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply service/cluster/xtra_db_cluster.go:32]: We already have another place where we have map like this. Let's postpone decision about it.

File: tests/xtradbcluster/xtra_db_cluster_test.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	"testing"
+
+	controllerv1beta1 "github.com/percona-platform/dbaas-api/gen/controller"
+	"github.com/percona-platform/dbaas-controller/tests"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestXtraDBClusterAPI(t *testing.T) {
+	kubeconfig := "todo" // TODO: set real kubeconfig
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply tests/xtradbcluster/xtra_db_cluster_test.go]: fixed

**Comment by BupycHuk** [comment]
@askomorokhov https://github.com/Percona-Lab/pmm-submodules/pull/1014

**Pull request merged**